File: pyZelda/entity.py
import pygame
from math import sin

class Entity(pygame.sprite.Sprite):
    def __init__(self, groups) -> None:
        super().__init__(groups)
        self.frame_index = 0
        self.animation_speed = 0.15
        self.direction = pygame.math.Vector2()

    def collision(self, direction):
        if direction == "horizontal":
            for sprite in self.obstacles_sprites:
                if sprite.hitbox.colliderect(self.hitbox):
                    if self.direction.x > 0:  # moving right
                        self.hitbox.right = sprite.hitbox.left
                    elif self.direction.x < 0:  # moving left
                        self.hitbox.left = sprite.hitbox.right
        elif direction == "vertical":
            for sprite in self.obstacles_sprites:
                if sprite.hitbox.colliderect(self.hitbox):
                    if self.direction.y > 0:  # moving down
                        self.hitbox.bottom = sprite.hitbox.top
                    elif self.direction.y < 0:  # moving up
                        self.hitbox.top = sprite.hitbox.bottom

    def move(self, speed):
        # make same speed in x and y direction
        if self.direction.magnitude() != 0:
            self.direction = self.direction.normalize()

        self.hitbox.x += self.direction.x * speed
        self.collision("horizontal")
        self.hitbox.y += self.direction.y * speed
        self.collision("vertical")

        self.rect.center = self.hitbox.center

    def wave_value(self):
        value = sin(pygame.time.get_ticks())
        if value > 0: return 255
        else: return 0

File: src/entity.cpp
#include "entity.hpp"
#include <iostream>
#include <sys/time.h>
#include <ctime>

Entity::Entity(const float &frame_index,
               const float &animation_speed,
               const py::Vector2f &direction,
               const std::list<std::shared_ptr<SpriteTexture>> &obstacle_sprites)
    : frame_index(frame_index),
      animation_speed(animation_speed),
      direction(direction),
      obstacle_sprites(obstacle_sprites)
{
}

int Entity::wave_value()
{
    struct timeval time_now;
    ::gettimeofday(&time_now, nullptr);
    time_t msecs_time = (time_now.tv_sec * 1000) + (time_now.tv_usec / 1000);

    auto value = sin(msecs_time);

    if (value > 0.f)
        return 255;
    else
        return 0;
}

void Entity::move(float speed)
{
    // move same speed in x and y direction
    if (this->direction.magnitude() != 0)
    {
        this->direction = this->direction.normalize();
    }

    this->hitbox_.transform(this->direction.x * speed, 0);
    this->collision("horizontal");
    this->hitbox_.transform(0, this->direction.y * speed);
    this->collision("vertical");

    this->rect_.center(this->hitbox_.center());
}

void Entity::collision(const std::string &direction)
{
    sf::Rect<float> intersection;
    if (!direction.compare("horizontal"))
    {
        for (const auto &sprite : this->obstacle_sprites)
        {
            if (sprite->colliderect(this->hitbox_, intersection))
            {
                if (this->direction.x > 0)
                { // move right
                    this->hitbox_.transform(intersection.width * -1, 0);
                }
                else if (this->direction.x < 0)
                { // move left
                    this->hitbox_.transform(intersection.width, 0);
                }
            }
        }
    }
    else if (!direction.compare("vertical"))
    {
        for (const auto &sprite : this->obstacle_sprites)
        {
            if (sprite->colliderect(this->hitbox_, intersection))
            {
                if (this->direction.y > 0)
                { // move down
                    this->hitbox_.transform(0, intersection.height * -1);
                }
                else if (this->direction.y < 0)
                { // move up
                    this->hitbox_.transform(0, intersection.height);
                }
            }
        }
    }
}
